### PR TITLE
finalizing scaling API work

### PR DESCRIPTION
**Pull request by cgbaker**
* adding raft and state_store support to track job scaling events
* updated the ScalingEvent API to record "message string,error bool" instead of cumbersome "reason,error *string"

### CODE REVIEW

File: api/scaling.go [hunk 1]
```diff
@@ -40,11 +40,11 @@ func (p *ScalingPolicy) Canonicalize(taskGroupCount int) {
...
 type ScalingRequest struct {
-	Count  *int64
-	Target map[string]string
-	Reason *string
-	Error  *string
-	Meta   map[string]interface{}
+	Count   *int64
+	Target  map[string]string
+	Message string
+	Error   bool
```
Comment (drewbailey) [inline api/scaling.go:46]: stylistic nit, Error being a bool seems unfamiliar, HasError might be a bit verbose :man_shrugging:

Comment (cgbaker) [reply api/scaling.go:46]: yeah, we played around with a few names there... it indicates that the autoscaler wants the Nomad operator to know that something is wrong. `HasError` might be better. but `Success/Failed` doesn't feel quite right. 

i wanted to use `ThereIsAProblemInTheAutoscalerOrTheScalingPolicy` but it seemed... not good.

File: nomad/job_endpoint.go [hunk 2]
```diff
@@ -896,79 +893,104 @@ func (j *Job) Scale(args *structs.JobScaleRequest, reply *structs.JobRegisterRes
...
+		reply.JobModifyIndex = job.ModifyIndex
 	}
 
-	// Populate the reply with job information
-	reply.JobModifyIndex = jobModifyIndex
-	reply.Index = reply.JobModifyIndex
-
-	// FINISH:
-	// register the scaling event to the scaling_event table, once that exists
+	// only create an eval for non-dispatch jobs and if the count was provided
```
Comment (drewbailey) [inline nomad/job_endpoint.go:937]:
```suggestion
	// Only create an eval for non-dispatch jobs and if the count was provided
```

File: nomad/state/state_store.go [hunk 3]
```diff
@@ -627,6 +627,85 @@ func (s *StateStore) DeleteDeployment(index uint64, deploymentIDs []string) erro
...
+	} else {
+		jobEvents = &structs.JobScalingEvents{
+			Namespace:     req.Namespace,
+			JobID:         req.JobID,
+			ScalingEvents: make(map[string][]*structs.ScalingEvent),
+		}
+	}
+
+	events := jobEvents.ScalingEvents[req.TaskGroup]
+	// prepend this latest event
```
Comment (drewbailey) [inline nomad/state/state_store.go:654]:
```suggestion
	// Prepend this latest event
```

File: nomad/state/state_store.go [hunk 4]
```diff
@@ -627,6 +627,85 @@ func (s *StateStore) DeleteDeployment(index uint64, deploymentIDs []string) erro
...
+	events = append(
+		[]*structs.ScalingEvent{req.ScalingEvent},
+		events...,
+	)
+	// truncate older events
```
Comment (drewbailey) [inline nomad/state/state_store.go:659]:
```suggestion
	// Truncate older events
```

File: api/scaling.go (hunk 1 above)
Comment (jazzyfresh) [reply api/scaling.go:46]: `ScalingRequest.HasError` and `ScalingRequest.Failed` both read well, and Failed has a precedent in other events structs

File: nomad/fsm.go [hunk 5]
```diff
@@ -51,6 +51,7 @@ const (
 	ScalingPolicySnapshot
 	CSIPluginSnapshot
 	CSIVolumeSnapshot
+	ScalingEventsSnapshot
```
Comment (jazzyfresh) [inline nomad/fsm.go:54]: smol smol ting, just double checking this snapshot case denotes multiple scaling events?

File: nomad/fsm.go [hunk 6]
```diff
@@ -1629,6 +1642,21 @@ func (n *nomadFSM) reconcileQueuedAllocations(index uint64) error {
 	return nil
 }
 
+func (n *nomadFSM) applyUpsertScalingEvent(buf []byte, index uint64) interface{} {
+	defer metrics.MeasureSince([]string{"nomad", "fsm", "upsert_scaling_event"}, time.Now())
+	var req structs.ScalingEventRequest
+	if err := structs.Decode(buf, &req); err != nil {
+		panic(fmt.Errorf("failed to decode request: %v", err))
```
Comment (jazzyfresh) [inline nomad/fsm.go:1649]: do we want to panic here? could we just fail the scaling event request and log the error?

File: nomad/fsm.go (hunk 5 above)
Comment (cgbaker) [reply nomad/fsm.go:54]: these snapshots are a single entry in the table. in this case, all of the scaling events for a single job.

File: nomad/fsm.go (hunk 6 above)
Comment (cgbaker) [reply nomad/fsm.go:1649]: this is the standard behavior for applying raft log entries to the state store. see, e.g.,: 
https://github.com/hashicorp/nomad/blob/88ff339f4d5bf461229e4bff018ef108fbb12d0a/nomad/fsm.go#L306-L311

it likely means that the raft log entry doesn't contain what the leading byte says it should contain, which likely means a version mismatch or corruption. the reason for the panic, i assume, is that state is undefined if we don't know how to apply a raft entry.

**Pull request merged**